Fix configure_onto when redis instance is early initialized

Drowze opened the pull request:
<!--
Check out the
https://github.com/DataDog/dd-trace-rb/blob/master/docs/DevelopmentGuide.md
for guidance on how to set up your development environment,
run the test suite, write new integrations, and more.
-->

**What does this PR do?**
Should fix #3277

**Motivation:**
Datadog.configure_onto doesn't work if the Redis instance was initialized before Datadog redis tracing is configured.

**Additional Notes:**
None.

**How to test the change?**
It is hard to automatically test this, but to manually test, simply use `Datadog.configure_onto` in a Redis instance that was initialized _before_ the Redis patch is applied (i.e. before the Datadog Redis tracing is configured).

**For Datadog employees:**
- [ ] If this PR touches code that signs or publishes builds or packages, or handles
credentials of any kind, I've requested a review from `@DataDog/security-design-and-guidance`.
- [ ] This PR doesn't touch any of that.

Drowze committed:
Fix configure_onto when redis instance is early initialized

Drowze commented on lib/datadog/tracing/contrib/redis/patcher.rb:
```diff
@@ -18,9 +18,15 @@ def self.included(base)
-                super(options.merge(redis_instance: self))
+              # we can't assume redis object is initialized after datadog is configured
+              def datadog_pin=(pin)
+                client_method = if Integration.redis_version >= Gem::Version.new('4.0')
+                                  :_client
+                                else 
+                                  :client
+                                end
```
Perhaps I should move this to to `integration.rb` ?

```ruby
# integration.rb
def client_method
  redis_version >= Gem::Version.new('4.0') ? :_client : :client
end

# patcher.rb
def datadog_pin=(pin)
  send(Integration.client_method).instance_variable_set(:@redis_instance, self)
  @datadog_pin = pin
end
```

Drowze replied:
Updated.

TonyCTHsu commented:
👋 Thanks @Drowze . I can reproduced the issue reported, it seems like Redis 3 and 4 are impacted. 

The fix seems feasible and I would like to move forward with test provided. 😄

TonyCTHsu committed:
Fail case for redis instance configuration

Drowze commented:
Hey @TonyCTHsu thank you very much for taking a care at the issue and even providing a test-case for it! (I wasn't sure about testing it as it was related to the patch lifecycle - looks like adding a new rake task was the way to go!)

Looking forward to v1.18.0 now 😄

Drowze commented on spec/datadog/tracing/contrib/redis/regression_spec.rb:
```diff
@@ -0,0 +1,49 @@
+require 'ddtrace'
+
+# The regression task must be a standalone task due to the life cycle of patcher
+RSpec.describe 'Regression', skip: Gem::Version.new(Redis::VERSION) >= Gem::Version.new('5') ? 'Not supported' : false do
```
Hey @marcotc 👋 
Thank you for your comment! It made me realize that my changes were NOT actually working for `configure_onto` for Redis >=5. Ops!
_(please note: no Rails involved here!)_

Following that, I just pushed a commit that:
- Removes mentions to `regression` - I agree this doesn't look like a regression!
- Make sure this test runs with Redis all redis supported versions (3, 4 and 5)
- And more importantly: makes configure_onto work both in Redis < 5 and Redis >=5 😄 

@marcotc @TonyCTHsu can you have another look please?

Drowze commented:
Just pushed a commit that:
- Removes mentions to `regression` - I agree this doesn't look like a regression!
- Make sure this test runs with Redis all redis supported versions (3, 4 and 5)
- And more importantly: makes configure_onto work both in Redis < 5 and Redis >=5 😄 

@marcotc @TonyCTHsu can you have another look please?

Drowze committed:
Add warning for usage of DD pin with redis >= 5

Also add multiple tests for custom tracing options at instance level:
- tests for custom tracing options with Datadog.configure_onto (for
  redis < 5)
- add tests for custom tracing options with redis-client's `custom`
  options (for redis >= 5)

Drowze commented on lib/datadog/tracing/contrib/redis/trace_middleware.rb:
```diff
@@ -71,10 +71,17 @@ def get_pipeline_commands(commands, command_args)
-
-            Datadog.configuration.tracing[:redis, redis_config.server_url].to_h.merge(custom)
+          def resolve(config)
+            # support both:
```
Hey @TonyCTHsu 👋 

Based on your reply, I understand that when using Redis 5+, the following no longer works (and is no longer supported):
```ruby
redis = Redis.new
Datadog.configure_onto(redis, service_name: 'my-redis')
```

I'm updating now the PR based on that :+1:

Drowze replied:
By the way, would we event want to support the following? (just wondering - I prefer to keep that out of the scope of this PR)
```ruby
require 'redis_client'

redis_client = RedisClient.new
Datadog.configure_onto(redis_client, ...)
redis_client.call("PING")
```

My feeling is that, since `configure_onto` is not supported on _redis_ 5+, it would be confusing to have it working with _redis-client_. Also because I believe redis-client can be initialized with custom options... i.e.: 
```ruby
RedisClient.new(custom_options: { datadog: { ... } })
```

Drowze commented:
Hey! 👋 

Sorry for the long wait to come back here - just updated this PR by:
- reverting the work bringing `Datadog.configure_onto` support to Redis 5+ (as we don't want to mess with that since `redis-client` has a supported API for defining configurations on custom middlewares)
- adding a change to trigger a warning everytime a user tries to use `Datadog.configure_onto` on a Redis instance 5+
    - note: added this as I think this may help informing users they're making wrong usage of ddtrace - which I believe may be a specially common scenario when the user might be migrating from Redis 4 to Redis 5
    - note2: the name of the module is a bit silly (`NotSupportedNoticePatch`) - definitely open to suggestions there 😄 
- adding/adapting tests related to using Redis custom tracing options on instance level (which is dependent on whether the Redis version is <5 or >=5)

I think the changes in this PR should make the differences between Redis 4 and Redis 5 tracing at instance level a bit more clear :)

Drowze committed:
Update gemfiles/*

TonyCTHsu committed:
Simplify pin logic

TonyCTHsu committed:
Add test for RedisClient with custom

TonyCTHsu committed:
Since the implementation changes without stale redis instance, remove standalone test tasks

TonyCTHsu committed:
Merge branch 'master' into fix-redis-configure_onto

TonyCTHsu commented:
👋 Thanks @Drowze , `NotSupportedNoticePatch` is a nice addition! 

I have carried on with the implementation, with some changes (I also merged `master` into this branch. )

1. Simply our pin logic during runtime so it no longer depends on stale redis instance. see [commit](https://github.com/DataDog/dd-trace-rb/pull/3278/commits/f4bb9fcbc3568caaa415d2e7a0092101d11a768a) 
2. Since the implementation no longer depends on stale redis instance, I believe we don't need a standalone life cycle test task
3. Add test for `RedisClient` for `custom` usage.

I wonder if you could remind me if I am missing anything? Would you be able to validate this change on your project to confirm the issue being fixed?

If yes, I would ask @marcotc for another round of review

TonyCTHsu committed:
Document standalone `RedisClient` usage

maycmlee commented on docs/GettingStarted.md:
````diff
@@ -1666,6 +1666,21 @@ customer_cache.get(...)
 invoice_cache.get(...)
 ```
 
+With standalone `RedisClient`
````
```suggestion
With a standalone `RedisClient`:
```

maycmlee reviewed (commented):
A couple of small suggestions!

marcotc committed:
Update docs/GettingStarted.md

Co-authored-by: May Lee <mayl@alumni.cmu.edu>

marcotc committed:
Sacrifice for the linting gods ☠️

TonyCTHsu commented:
👋 @Drowze , `1.20.0` has been released! Give it a try 👍
